src/pandas_profiling/model/dataframe_wrappers.py

```diff
@@ -0,0 +1,599 @@
+
+        Args:
+            n: top n rows
+
+        Returns: dataframe with only top n rows
+
+        """
+        pass
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Get the number of rows in a dataframe as an int
+
+        Returns: number of rows in column
+
+        """
+        pass
+
+    @abstractmethod
+    def __getitem__(self, key):
+        pass
+
+
+class PandasDataFrame(GenericDataFrame):
+    """
+    This class is the abstract layer for the pandas dataframe.
+
+    It implements the check_if_corresponding_engine function to check if an obj is a pandas dataframe and can be wrapped.
+
+    It implements the get_pandas_df method to enable singledispatch wrapped functions to
+    operate on the internal pandas dataframe without breaking abstraction
+
+    """
+
+    engine = "pandas"
+
+    def __init__(self, df):
+        super().__init__()
+        # self.df holds the underlying data object
+        self.df = df
+        self.n_rows = len(self.df)
+
+    @staticmethod
+    def check_if_corresponding_engine(obj) -> bool:
+        """
+        Check if pandas dataframe using isinstance. More pythonic way of checking as opposed to spark type check.
+
+        Args:
+            obj: check if object is a pandas dataframe
+
+        Returns: True if the __module__ and __name__ of object matches pandas dataframe, else false
+        """
+        return isinstance(obj, pd.DataFrame)
+
+    @staticmethod
+    def preprocess(df: pd.DataFrame) -> pd.DataFrame:
+        """Preprocess the dataframe
+
+        - Appends the index to the dataframe when it contains information
+        - Rename the "index" column to "df_index", if exists
+        - Convert the DataFrame's columns to str
+
+        Args:
+            df: the pandas DataFrame
+
+        Returns:
+            The preprocessed DataFrame
+        """
+
+        # Treat index as any other column
+        if (
+            not pd.Index(np.arange(0, len(df))).equals(df.index)
+            or df.index.dtype != np.int64
+        ):
+            df = df.reset_index()
+
+        # Rename reserved column names
+        df = rename_index(df)
+
+        # Ensure that columns are strings
+        df.columns = df.columns.astype("str")
+        return df
+
+    @property
+    def columns(self) -> List[str]:
+        return self.df.columns
+
+    @property
+    def empty(self) -> bool:
+        return self.df.empty
+
+    def get_duplicate_rows_count(self, subset) -> int:
+        """
+
+        Args:
+            subset:
+            keep:
+
+        Returns:
+
+        """
+        return len(self.df[self.df.duplicated(subset=subset, keep="first")])
+
+    def dropna(self, subset) -> "PandasDataFrame":
+        return PandasDataFrame(self.df.dropna(subset=subset))
+
+    def groupby_get_n_largest_dups(
+        self, supported_columns, n_head
+    ) -> Tuple[Dict[str, Any], Optional[pd.DataFrame]]:
+
+        metrics: Dict[str, Any] = {}
+        if n_head > 0:
```

Comment:
I would consider re-writing this function just to reduce the level of nesting e.g.

```
if n_head <= 0:
  return metrics, None

if not supported_columns or len(self.df) == 0:
  metrics["n_duplicates"] = 0
  metrics["p_duplicates"] = 0.0
  return metrics, None

// remainder of un-nested implementation
```

Reply:
Also, that can just be `self.df.empty`